src/main/java/de/tum/in/www1/artemis/config/SAML2Configuration.java

```diff
@@ -132,8 +132,7 @@ private static X509Certificate readPublicCert(String file) throws IOException, C
 
     private RSAPrivateKey readPrivateKey(String file) throws IOException {
         // Read PKCS#8 File!
-        try (FileReader keyReader = new FileReader(file, StandardCharsets.UTF_8)) {
-            PEMParser pemParser = new PEMParser(keyReader);
+        try (PEMParser pemParser = new PEMParser(new FileReader(file, StandardCharsets.UTF_8))) {
```

Comment:
Just one question where I’m not sure about the Java semantics:

You now combine the two AutoCloseables so that one is given as argument to the other. Does Java still track the inner `FileReader` to auto-close it, or would you have to declare them as `try (var fr = new FileReader(…); var pemParser = new PemParser(fr)) { … }` to make sure both are closed properly at the end?

Reply:
Yes, Java does not that automatically (and therefore you cannot see that in you example), but this is the case for BufferedReader and BufferedWriter:
```
 public void close() throws IOException {
        synchronized (lock) {
            if (out == null) {
                return;
            }
            try (Writer w = out) {
                flushBuffer();
            } finally {
                out = null;
                cb = null;
            }
        }
    }
```

The PEMParser is a subclass of BufferedReader that also uses this implementation, you can test this:
```
public static void main(String[] args) {
        try (PEMParser parser = new PEMParser(new Closing())) {

        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
```

prints `Closing`.

Reply:
Thanks, at that point you are relying on an internal implementation detail, though. For me personally that is a very big red flag and should definitely be avoided.

Reply:
Yes it's an implementation detail, but it's consistent for all Java writers and readers. So I'd say that that's ok.

Reply:
I definitely agree @b-fein we should not rely on internals.

Nevertheless, in this specific case it is not an internal behavior I guess.
I know that readers guarantee that they close the associated steams as well : https://docs.oracle.com/javase/8/docs/api/java/io/Reader.html

I think therefore we may can go ahead with the change :)

Reply:
Maybe @Strohgelaender you can have a short look whether the behavior is covered by the documentation of close for the classes.

Reply:
I now changed this case to two statements. The `PEMParser` is part of another external library so here we should stay safe and not rely on internal implementations. I think it's another thing for classes from the java API, so I kept it there.